src/boot/efi/stub.c

```diff
@@ -159,23 +162,76 @@ static void export_stub_variables(EFI_LOADED_IMAGE_PROTOCOL *loaded_image) {
         (void) efivar_set_str16(MAKE_GUID_PTR(LOADER), u"StubInfo", u"systemd-stub " GIT_VERSION, 0);
 
         (void) efivar_set_uint64_le(MAKE_GUID_PTR(LOADER), u"StubFeatures", stub_features, 0);
+
+        (void) efivar_set_uint64_str16(MAKE_GUID_PTR(LOADER), u"StubProfile", profile, 0);
+}
+
+static bool parse_profile_from_cmdline(char16_t **cmdline, unsigned *ret_profile) {
+        assert(cmdline);
+        assert(*cmdline);
+        assert(ret_profile);
+
+        const char16_t *p = *cmdline;
+        if (p[0] != '@')
+                goto nothing;
+
+        uint64_t u;
+        const char16_t *tail;
+        if (!parse_number16(p + 1, &u, &tail))
+                goto nothing;
+        if (u > UINT_MAX)
+                goto nothing;
+        /* Remove exactly one separating space. No further mangling, in order to not disturb measurements –
+         * and thus making prediction harder –, after all we want that people can safely prefix their command
+         * lines with a profile without having to be bothered with additional whitespace the command line
+         * might already contain. */
+        if (tail[0] == u' ')
+                tail++;
+        else if (tail[0] != 0) /* If this is neither a space nor the end of the string, it must be something else */
+                goto nothing;
+
+        /* Drop prefix */
+        free_and_xstrdup16(cmdline, tail);
+        *ret_profile = u;
+        return true;
+
+nothing:
+        *ret_profile = 0;
+        return false;
+}
+
+static bool parse_profile_from_argument(const char16_t *arg, unsigned *ret_profile) {
+        assert(arg);
+        assert(ret_profile);
+
+        if (arg[0] != '@')
+                goto nothing;
+
+        uint64_t u;
+        if (!parse_number16(arg + 1, &u, /* ret_tail= */ NULL))
+                goto nothing;
+
+        if (u > UINT_MAX)
+                goto nothing;
+
+        *ret_profile = u;
+        return true;
+
+nothing:
+        *ret_profile = 0;
+        return false;
 }
 
-static bool use_load_options(
+static void process_arguments(
                 EFI_HANDLE stub_image,
                 EFI_LOADED_IMAGE_PROTOCOL *loaded_image,
-                bool have_cmdline,
-                char16_t **ret) {
+                unsigned *ret_profile,
+                char16_t **ret_cmdline) {
 
         assert(stub_image);
         assert(loaded_image);
-        assert(ret);
-
-        /* We only allow custom command lines if we aren't in secure boot or if no cmdline was baked into
-         * the stub image.
-         * We also don't allow it if we are in confidential vms and secureboot is on. */
-        if (secure_boot_enabled() && (have_cmdline || is_confidential_vm()))
-                return false;
+        assert(ret_profile);
+        assert(ret_cmdline);
 
```

Comment:
I'm pretty sure we should keep this in `process_arguments()`. We shouldn't be trying to look at potentially untrusted stuff when we're in secure boot mode or in a confidential VM

Reply:
this check is not dropped. it's just moved into a separate function settle_command_line(), which deals with this.

I moved this because the idea of this multi-profile work is that we have multiple valid profiles the user can pick from, even in SB mode. But that means we *must* look at the cmdline to parse out the profile id, and then store the rest but not use it until we came to the conclusion that that's safe. And that judgement we make in settle_command_line().